### ISSUE TITLE
Typesafe API to retrieve enums from strings

### ISSUE THREAD

**Issue by dimo414** [timostamm/protobuf-ts#642]
To my knowledge, TypeScript's support for going [from a string to an enum](https://github.com/timostamm/protobuf-ts/blob/main/MANUAL.md#enum-representation) requires that the string be typed as `keyof typeof MyEnum`. This works for string literals like the example but doesn't work for arbitrary strings which could be invalid values. Here's a [playground demo](https://www.typescriptlang.org/play?#code/KYOwrgtgBAsgngUXNA3gKCpqBBAcgTSgF4oAGAGgy3wQGVioBGSrKXAeQYCY0BfNNAGMA9iADOwgDbAAdJOEBzABTwkkANoAiGrU0BdAJQBuNNIAuUANbA4DTTGz4AQgk0mR4qbPnLVyddZwhkZAA) which reports an `Element implicitly has an 'any' type` error.

As a more concrete example, I am receiving a string from an untyped client that "should" be an enum value. To satisfy the type checker I need to cast the key `as` the expected type, and then handle the potential `undefined` that is returned on a miss to convert it to the enum's 0th value:

```
let key = "MAYBE";
// notice the type here is "wrong" because the index lookup returns undefined,
// but I don't want to pollute myEnum with |undefined, I want it to use the default
let myEnum: MyEnum = MyEnum[key as keyof typeof MyEnum];
if (myEnum === undefined) {
  myEnum = MyEnum.ANY;
}
```

It would be really useful if protobuf-ts provided a more ergonomic string-to-enum utility than what [TypeScript provides](https://www.typescriptlang.org/docs/handbook/enums.html#enums-at-compile-time). At a minimum it should support arbitrary strings without needing casting, and potentially it could behave like the protobuf spec and return the 0th enum value rather than `undefined`.

**Comment by jcready** [comment]
You should be able to create these functions yourself ([playground](https://www.typescriptlang.org/play?#code/KYOwrgtgBAsgngUXNA3gKCpqBBAcgTSgF4oAGAGgy3wQGVioBGSrKXAeQYCY0BfNNADMwIAMYAXAJYB7EFEkBnANLA40wQB4AKlGAAPcaAAmCqCl4A+ABTiAXFC3koAa1X2F4gE6SQAcwCU9q5w8qYe3n5QAGQuquoOZlSYnsDiYJ5ywfJy4gDcfALCYlKyUL6pAAqe0uLSSJAAagCGADZgwOyeACLAgk1gLeLaugbGpubWwPaOsXDuXj4B9qRQAD4OANrhi9Gz8VoAuomsKWkZoSpqglbATsH+UAD8uhvBR8v5-GiisgrSLcAAHQtaS+KzweoQDYAIhotGhB38+QB4lmDGhMGw+AAQghoflvr9UU0GOVxFUanVkM02h1ur1+oNwYhkHdVEi0AB6TmsAB6j0JIA8UAARgwIcgNk0DvluXyBWhJIIoFZFJd1MzIWy4P4HuhWD8hf8gSCwRLIK9VIjPkA)):

```ts
function isKeyof<T extends {}>(t: T, key: string): key is string & keyof T {
    return key in t;
}

function getEnumValueOrDefault<T extends {}>(e: T, key: string): 0 | T[string & keyof T] {
    return isKeyof(e, key) ? e[key] : 0;
}
```

**Comment by dimo414** [comment]
Yeah I agree it's possible to workaround, but I was hoping protobuf-ts could include these e.g. in the generated code of each enum so it's right on the type, rather than something we need to pull together or go find.

**Comment by jcready** [comment]
The enums are just TS enums. No extra properties or methods can be added and have them be TS enums.

**Comment by dimo414** [comment]
Fair enough; well IMO there's room for protobuf-ts to improve the ergonomics of string->enum lookups vs. what TypeScript provides out of the box, but I understand if you disagree.
